src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/LengthBucketsFrozenDictionary.cs

```diff
@@ -50,59 +48,59 @@ internal sealed class LengthBucketsFrozenDictionary<TValue> : FrozenDictionary<s
                 return null;
             }
 
-            // Iterate through all of the inputs, bucketing them based on the length of the string.
-            var groupedByLength = new Dictionary<int, List<KeyValuePair<string, TValue>>>();
-            for (int i = 0; i < keys.Length; i++)
-            {
-                string s = keys[i];
-                Debug.Assert(s.Length >= minLength && s.Length <= maxLength);
-
+            int arraySize = spread * MaxPerLength;
 #if NET6_0_OR_GREATER
-                List<KeyValuePair<string, TValue>> list = CollectionsMarshal.GetValueRefOrAddDefault(groupedByLength, s.Length, out _) ??= new List<KeyValuePair<string, TValue>>(MaxPerLength);
+            if (arraySize >= Array.MaxLength)
 #else
-                if (!groupedByLength.TryGetValue(s.Length, out List<KeyValuePair<string, TValue>>? list))
-                {
-                    groupedByLength[s.Length] = list = new List<KeyValuePair<string, TValue>>(MaxPerLength);
-                }
+            if (arraySize >= 0X7FFFFFC7)
 #endif
-
-                // If we've already hit the max per-bucket limit, bail.
-                if (list.Count == MaxPerLength)
-                {
-                    return null;
-                }
-
-                list.Add(new KeyValuePair<string, TValue>(s, values[i]));
-            }
-
-            // If there would be too much empty space in the lookup array, bail.
-            if (groupedByLength.Count / (double)spread < EmptyLengthsRatio)
             {
+                // This size check prevents OOM.
```

Comment:
Not really :) It just prevents artificial OOMs from arrays of unsupported sizes.  You could still try to rent an array of length Array.MaxLength - 1 and OOM.